### PR TITLE
[usage]Actually reset usage during ResetUsage RPC

### PR DESCRIPTION
**Pull request by easyCZ**
## Description

Reset during ResetUsage RPC.
Also resets usage when fetching a CostCenter, in case it has just expired.

## Related Issue(s)

* Part of https://github.com/gitpod-io/gitpod/issues/14178

## How to test

Unit tests

## Release Notes

```release-note
NONE
```

## Werft options:

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [ ] /werft with-preview
- [ ] /werft with-large-vm
- [ ] /werft with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`

/hold

### CODE REVIEW

File: components/usage/pkg/db/cost_center.go
Comment (svenefftinge) [inline components/usage/pkg/db/cost_center.go:267]: we are doing most of the logic in `upgradeToStripe`, which is ill-named as it doesn't to anything specific to stripe but resets the usage. Would be great if you merge those functions.

File: components/usage/pkg/db/cost_center.go
Comment (svenefftinge) [inline components/usage/pkg/db/cost_center.go:301]: why do you touch the spending limit here? shouldn't we just set `cc.SpendingLimit`?

File: components/usage/pkg/db/cost_center.go
Comment (svenefftinge) [inline components/usage/pkg/db/cost_center.go:293]: It also needs to be stored. `upgradeToStripe` does this and is ill-named (doesn't do any stripe specific things). Would be good if you can rename and reuse or so

File: components/usage/pkg/db/cost_center.go
Comment (svenefftinge) [inline components/usage/pkg/db/cost_center.go:282]: no need to obtain the default spending limits as the cost center already has one

File: components/usage/pkg/db/cost_center.go
Comment (easyCZ) [reply components/usage/pkg/db/cost_center.go:301]: I did consider just carrying it over, but at the moment, there isn't a use-case for keeping the SpendingLimit the same as before (such that it would be different to the defaults). This ensures that it does reset it to the defaults (and sets any updated defaults). We've already forgotten to configure the spending limit defaults so existing cost centers had 1B. This would've ensured that it does reset after we do update the default spending limit.

We can change it, but we'll need to manually go and fix the 1B for all individual users, and whenever we change the default spending limit (and hence free credit allowance), we'll also need to do a bulk update.

File: components/usage/pkg/db/cost_center.go
Comment (easyCZ) [reply components/usage/pkg/db/cost_center.go:282]: See https://github.com/gitpod-io/gitpod/pull/14249#discussion_r1007691049 for reasoning.

File: components/usage/pkg/db/cost_center.go
Comment (easyCZ) [reply components/usage/pkg/db/cost_center.go:293]: Good spot. I'll rename the methods because `Compute` suggests it does more than just a create which doesn't align with the Go principle of communicating work/complexity.

File: components/usage/pkg/db/cost_center.go
Comment (svenefftinge) [reply components/usage/pkg/db/cost_center.go:301]: Ah, that's why you do that. Sounds ok to me. Could you add a comment for the reader?

File: components/usage/pkg/db/cost_center.go
Comment (svenefftinge) [reply components/usage/pkg/db/cost_center.go]: Fine with renaming. FWIW I usually use calculate or compute to indicate that a function doesn't do any side-effects such as persisting in DB. Does calculate imply persistence fir you?

Comment (easyCZ) [reply components/usage/pkg/db/cost_center.go]: To some degree. I guess normally I'd be "looking out" for the "New" keyword which mostly signals it's just a constructor and not an action  but this depends.

In any case, I should've checked regardless but I've been short for time and wanted to put this up in case the team needs to be unblocked.

File: components/usage/pkg/db/cost_center.go
Comment (easyCZ) [reply components/usage/pkg/db/cost_center.go:305]: Done.

**Comment by easyCZ** [comment]
Thanks for the review @svenefftinge. There are still unit tests missing for this PR before we can land it.

**Comment by geropl** [comment]
Starting review now.

File: components/usage/pkg/db/cost_center.go
Comment (geropl) [inline components/usage/pkg/db/cost_center.go:314]: Do I get this right: If the previous billing cycle ended already, we start the new one _now_ instead of based on the old one + iterating months until now?

Just asking because conceptually, the later might be simpler to grasp mentally and thus might be easier to debug and change later on.

Comment (easyCZ) [reply components/usage/pkg/db/cost_center.go:314]: This is just a safety net for the case where no NextBillingTime was present on the record. If it is present, we do (old.NextBillingTime + 1 month). Otherwise, we do (now + 1 month)

**Review by geropl — APPROVED** [review]
Tests pass, and code LGTM

/hold for [this comment](https://github.com/gitpod-io/gitpod/pull/14249/files#r1011333934) which is about sth conceptually where I might lag overview, so please decide how to handle it. :pray:

File: components/usage/pkg/db/cost_center.go
Comment (geropl) [reply components/usage/pkg/db/cost_center.go:314]: > f it is present, we do (old.NextBillingTime + 1 month). Otherwise, we do (now + 1 month)

And if there is a gap of say 3 months? Where do we go from old.nextBillingTime to sth that covers today? :thinking:

Comment (easyCZ) [reply components/usage/pkg/db/cost_center.go:314]: There's a periodic job which triggers the reset for expired every 15 minutes. This should cover most cases.

In the case it the CostCenter expires between now and (-15 minutes), we reset when we're fetching the CostCenter.

In the rare case where we've been failing to reset for a while, the reset catches up by incrementally (in each iteration) moving the BillingTime forward by 1 month.

Comment (geropl) [reply components/usage/pkg/db/cost_center.go:314]: > There's a periodic job which triggers the reset for expired every 15 minutes.

Great, solved. :heavy_check_mark: :lotus_position:

**Pull request merged**
